### ISSUE TITLE
feature: design publish controls 

### ISSUE THREAD

**Issue by jku** [theupdateframework/tuf-on-ci#54]
There is likely no perfect publish workflow for everyone but we should have something that we imagine would work for a lot of users (and the rest can invent their own solution).

Some options 
1. automated publishing (with a automated staging phase) but the online role changes lead to a PR (and not automatic changes): this means all metadata changes are gated behind maintainer approval so publishing could likely be automated
2. automated online metadata changes, but publishing gated behind maintainer approval
3. automated online metadata changes, automated publishing to a "staging/testing/unstable" url (I'll call it _testing_ from now on). Promoting _testing_ to _production_ is gated on maintainer approval

I think I like option 3 -- It allows for some repositories to operate without maintainer gating (_testing_ is then the production repository) but also allows full control of release to production

**Comment by jku** [comment]
The way I imagine staging/testing/unstable to work (with some handwaving WRT how much GitHub Actions actually supports):
* metadata changes are published to staging/testing/unstable url ASAP
* automated tests that run a downloader client that uses staging/testing/unstable: their results are made available, failures lead to issues filed
* optional for projects with production promotion gating: 
  * maintainers are made aware of upcoming production promotion
  * project has a production promotion schedule: maintainers approve the promotion x days after it appears in staging/testing/unstable (with exceptions for security releases etc)

**Comment by jku** [comment]
I guess the more traditional alternative (from a code review perspective) would be to review all timestamp/snapshot changes as PRs:
![tuf-on-ci-publish-alt](https://github.com/theupdateframework/tuf-on-ci/assets/31889/43fcb744-7834-49e8-8ff1-fcc9e3a9fb76)

Here the maintainer approval is the PR review: 
* all security checks should still be done before online roles sign
* the pr publish should be made available to clients
* PR review can then include automated client tests and even maintainers running a client manually
* when timestamp/snapshpot PR is merged, the production publish happens automatically

In this model we do not get a single "testing repository URL": testing requires configuring the client for this specific testing repository

TUF-on-CI could still have a config that allows automatic merges for the projects that don't want to constantly review timestamp PRs?

**Comment by kommendorkapten** [comment]
Another option I discussed today with @jku is the availability to verify current metadata as part of an online signing. This is more to make sure that the `main` branch is in a good state. 

As an example, during any online (snapshot or timestamp) signing, the following would be verified:
1. `root.json` is consistent WRT the historical metadata in `metadata/root_history`
2. `targets.json` is signed by expected keys as defined in `root.json`
3. All targets in `targets/` directory matches whats defined in `targets.json`
4. Online roles (snapshot and timestamp) are signed by the expected key specified in `root.json`

If the state is not expected, the automation should create an issue to notify the repository maintainers that `main` branch needs fixes. If the state is not as expected, the publish should proceed.

**Comment by jku** [comment]
I've had trouble figuring out how to handle the different cases in a single flow  but I think this is the simplest I've managed to make it:
![online-sign](https://github.com/theupdateframework/tuf-on-ci/assets/31889/54c31d3d-0c01-414f-92e9-fd4e7b97222a)

Basically, we can have a single point in the workflow where both
* online signing may happen
* publish may happen

and then these two don't need to happen anywhere else. The critical edge cases are:
* metadata changes might not lead to online signing but should always lead to publish
* if the online signing is triggered by cron, then "no-op" is a possible result (no online signing needed, no publish needed)

**Comment by jku** [comment]
Playing with this more, I think I want to tweak the whole stack from workfows to actions to python code a bit.

* We should remove current snapshot and version-bumps workflows: I think I chose the  wrong abstractions there
* Instead have 
  * `online-sign` (invoked by cron and push to main) 
  * `create-signing-event` (invoked by cron)
* This is more logical, removes the duplication in calling online-version-bump and snapshot,  and better matches the GitHub permissions required as well

`create-signing-event` would be like the offline-version-bump part of current `version-bumps`  workflow: It runs on cron and creates new signing events (branches and issues) for any roles that are in signing period and do not have a signing event yet

`online-sign` is the bubble in the diagram above: it runs on cron but also on push to main. 
* invoke the `snapshot` action (which we could also rename `online-sign`). snapshot 
* if running on cron and new snapshot/timestamp did not happen, exit early successfully (no need to create a repository version). Otherwise create a new repository version (a GH artifact)
* publish the repository version artifact


The big change here would be in actions: `online-version-bump` action is no longer needed as the snapshot action handles that too -- the only thing that really needs to change is that we need  a way to to tell the action that sometimes a repository version artifact is needed even if snapshot is not created (this is after a signing event)

**Comment by jku** [comment]
Drafting this a bit further:

![online-sign](https://github.com/theupdateframework/tuf-on-ci/assets/31889/854a98e3-d728-4b64-af4d-32251e17f4d3)


## Actions

The new actions replace current `online-version-bump` and `snapshot`.

**online-sign**

* has argument `force-upload`, this is used when a new repository release is needed even when online roles do not change (in practice used when push triggered on cases like only root metadata changing)
* Creates and signs a new timestamp if needed, creates a new snapshot if needed. Commits these to main
* Uploads a new repository release artifact if new metadata was created or if `force-upload`

**publish**

* downloads repository release artifact
* Publishes the repository release artifact to Github Pages

**create-signing-events**

* renamed from offline-version-bumps to be more consistent

## Workflows

The new workflows replace current `snapshot` and `version-bumps`.

**online-sign-publish**

* is part of the user repository, originally comes from tuf-on-ci-template
* has triggers for both push and cron
* invokes the `online-sign` action from tuf-on-ci (with arg to force repo version upload)
* if a repository version was created, invokes `publish` action

**create-signing-events**

* is part of the user repository, originally comes from tuf-on-ci-template
* has a cron trigger
* invokes `create-signing-events` (which is just the existing `offline-version-bump`)


## What practically changes

The current actions workflows have inconsistent responsibilities:
* actions online-version-bump and snapshot do mostly the same thing
* workflow snapshot and half of version-bumps are the same thing

the changes fix this:
* there's just one action and one workflow that do online changes
* signing event creation (for roles that are in signing period) is not bundled with online version bumping

**Comment by jku** [comment]
One more possible improvement: I think the idea in previous comment is correct but still does not separate publishing from online signing enough. Imagine a case where the published repository hosting is lost and we need to publish the current version right now:
* we can't just publish from main because it's not guaranteed to be publishable
* there is no tag or branch that is known publishable

so the added tweak is:
* online sign does not actually publish, it documents the current publishable point by pushing to branch `publish`: this branch always points to last "good" repository version
* push to `publish` triggers a separate publishing workflow
* now the solution to the imagined problem (where a re-publish is needed) is to manually trigger the publishing workflow from `publish` branch 

![online-sign](https://github.com/theupdateframework/tuf-on-ci/assets/31889/3d56e45b-b6f2-4a1f-bc13-2b61f5cc7e98)

**Comment by jku** [comment]
status update: have WIP versions of the two last "paths" in the diagram (_scheduled signing event creation_ and  _publish_). So far I think it looks decent. Next I will have a look at implementing online-sign.

I'm starting with the assumption that this gets tested all at once: "API breaks" in workflows/actions are annoying enough that I'd rather only do a single one even for testing.

**Comment by kommendorkapten** [comment]
Nice progress! Let me know if you need some help testing this out.

**Comment by jku** [comment]
It's very close now: I'll probably ping you today that it's good to test. Could plan to have a meeting on monday to discuss the decisions I made.

Documenting some mention worthy things here already so I remember:
* Amount of code is roughly the same but I think both actions and workflows are much easier to understand now
* online-sign workflow moving the publish branch looks pretty good but has consequences:
  * pushing to a branch from a workflow does not trigger a push event so that means online-sign needs to explicitly dispatch the publish workflow (I still prefer to keep online-sign and publish separate since publish is something users should be able to modify)
  * GH pages by default allows deployments from "main", not "publish" branch : this means one more required configuration thing during setup (or alternatively we rearrange our branch naming even more)
* online-sign now produces GH artifacts named "metadata" and "artifacts" because that makes a lot more sense... I did decide to leave the github Pages specific artifact in there as well still: this is just an optimization to make the default gh pages publish action simple as possible 
* More sensible workflows means that sigstore online signing now works with a single key: this is great but is a very breaking change for anyone using sigstore online signing (requires a signing event when upgrading)

**Comment by jku** [comment]
Note on the about PR: it doesn't introduce actual publishing controls but it makes it possible for individual projects to easily modify their publishing workflow.
